plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+	}
+	return nil
+}
+
+// accumulateServerDiagnostics accumulates statistics from the compute(nova) service.
+// currently only supports 'libvirt' driver.
+func (o *OpenStack) accumulateServerDiagnostics(acc telegraf.Accumulator) {
+
+	var type_port = regexp.MustCompile(`_rx$|_rx_drop$|_rx_errors$|_rx_packets$|_tx$|_tx_drop$|_tx_errors$|_tx_packets$`)
+	var type_cpu = regexp.MustCompile(`cpu[0-9]{1,2}_time$`)
+
+	for server_id, diagnostic := range o.diag {
+
+		tags := map[string]string{
+			"server_id": server_id,
+		}
+		fields := map[string]interface{}{}
+
+		port_name := make(map[string]bool)
+
+		// for metrics other than port/interface or cpu
+		other_metrics := make(map[string]interface{})
+		s, ok := diagnostic.(map[string]interface{})
+		if !ok {
+			o.Log.Warnf("unknown type for diagnostics %T", diagnostic)
+			continue
+		}
+
+		for k, v := range s {
+			if type_port.MatchString(k) {
+				port_name[strings.Split(k, "_")[0]] = true
+			} else if type_cpu.MatchString(k) {
+				fields[k] = v
+			} else {
+				other_metrics[k] = v
+			}
+		}
+
+		fields["hdd_errors"] = other_metrics["hdd_errors"]
+		fields["hdd_read"] = other_metrics["hdd_read"]
+		fields["hdd_read_req"] = other_metrics["hdd_read_req"]
+		fields["hdd_write"] = other_metrics["hdd_write"]
+		fields["hdd_write_req"] = other_metrics["hdd_write_req"]
+		fields["memory"] = other_metrics["memory"]
+		fields["memory-actual"] = other_metrics["memory-actual"]
+		fields["memory-rss"] = other_metrics["memory-rss"]
+		fields["memory-swap_in"] = other_metrics["memory-swap_in"]
+		fields["vda_errors"] = other_metrics["vda_errors"]
+		fields["vda_read"] = other_metrics["vda_read"]
+		fields["vda_read_req"] = other_metrics["vda_read_req"]
+		fields["vda_write"] = other_metrics["vda_write"]
+		fields["vda_write_req"] = other_metrics["vda_write_req"]
+		tags["no_of_ports"] = strconv.Itoa(len(port_name))
+
+		if len(port_name) > 0 {
+			for key := range port_name {
+				fields["port_rx"] = s[key+"_rx"]
+				fields["port_rx_drop"] = s[key+"_rx_drop"]
+				fields["port_rx_errors"] = s[key+"_rx_errors"]
+				fields["port_rx_packets"] = s[key+"_rx_packets"]
+				fields["port_tx"] = s[key+"_tx"]
+				fields["port_tx_drop"] = s[key+"_tx_drop"]
+				fields["port_tx_errors"] = s[key+"_tx_errors"]
+				fields["port_tx_packets"] = s[key+"_tx_packets"]
+				tags["port_name"] = key
+				acc.AddFields("openstack_server_diagnostics", fields, tags)
+			}
+		} else {
+			acc.AddFields("openstack_server_diagnostics", fields, tags)
+		}
+	}
+}
+
+// accumulateStacks accumulates statistics from the stack service.
+func (o *OpenStack) accumulateStacks(acc telegraf.Accumulator) {
+	for _, stack := range o.stacks {
+		tags := map[string]string{
+			"creation_time": stack.CreationTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
+			"description":   stack.Description,
+			"id":            stack.ID,
+			"name":          stack.Name,
+			"stack_tags":    strings.Join(stack.Tags[:], ","),
```

Comment:
Actually we wanted to use them directly but the tags in Openstack are of array format ([ ]string) where as in Telegraf it is map.
So, is there any way we can use actual Openstack tags directly? 

We thought of two ways,
1. to use static names as keys in map with tags as values(eg: "tag_1": "prod"), this will also be of not much use because of uncertainty of where the tag would be placed (tag_n).
2. to use tag values as keys and set value to "true" (eg: "prod": "true"), with this approach the cardinality might increase drastically if there are large number of unique tags in Openstack for a particular service type.

Still think option 2 is worth incorporating but wasn't sure if the Influx community will agree. What do you think?

Reply:
Oh I see. Didn't notice that when making the comment. I agree that option 2 is worth a thought. My preferred way to handle cardinality issues would be to use the telegraf `taginclude`/`tagexclude` mechanisms. Let me discuss this with the core-developers and I'll get back to you on this issue. Ok?

Reply:
That's sounds good.

Reply:
Any update on this topic?

Reply:
Hey @singamSrikar, sorry for this taking longer than expected. I talked to @reimda and we think that adding the tags as separate metric tags is the best way. The idea is to allow the user to specify a `tag_prefix` e.g. `tag_prefix = "openstack_tag_"` (which could be the default). Furthermore, let the user specify a `tag_value`, as Openstack tags do not have a value. The default could be empty `tag_value = ""`. 

We then convert the OpenStack tags e.g. `red,blue,green` into metric tags. With the default values for `tag_prefix` and `tag_value` this will result in `openstack_tag_red=""`, `openstack_tag_green=""`, `openstack_tag_blue=""` on the metric.
This gives the user maximum flexibility with sensible defaults. We could even disable those tags by setting empty `tag_prefix=""` .

What do you think?

Reply:
Hey @srebhan, I do not think that would a good idea as 1)  tags `red,blue,green` would create 3 new columns in the measurement and similarly there could be n different tags leading to n different columns increasing cardinality drastically (for eg:. hypervisor has about 30 tags), and 2) with tag value as empty string we would not be able to determine if a particular entry in the measurement has this tag or not.

Reply:
@singamSrikar I share your concerns about cardinality, but let me clarify things

> 1) tags red,blue,green would create 3 new columns in the measurement and similarly there could be n different tags leading to n different columns increasing cardinality drastically (for eg:. hypervisor has about 30 tags)

For influxdb the additional tags are **not** a big issue as they are only stored once per series. The _space_ increase is negligible. However, the combinatorics is not, i.e. if you always have the same 30 or 1000 tags you only have one series and we only store this tag-set _once_ per series **not** once per datapoint.

> with tag value as empty string we would not be able to determine if a particular entry in the measurement has this tag or not.

You might be right here. We can circumvent it by setting a defined (maybe even user-defined) value.

Let me additionally comment on cardinality. In your current scheme, you will produce at least the same cardinality compared to separating out the tags. The cardinality is defined by the measurement _name_, the field _name_ and the tag **set** (consisting of tag _name_ **and** _value_). By joining the Openstack tags you create a certain metric tag _value_ resulting in a specific series. Even worse, if you do not guarantee the same tag order for each datapoint when joining you might produce __different__ series for the same tag-set! In the end the cardinality when joining data the way you do is the same as if you have separate tags. And worse...

If you join the Openinput tags, you basically render them useless. Users cannot query them (that's what tags are for in the first place), but the are degraded to simple display values. Imagine you have multiple stacks/hypervisors a first one with the tags `production, region-eu, customer_x` and another one with `region-us, customer_y, ha, production, priority_high` in you schema you will get two series, the same for my schema. However, I can then craft a query returning all `production` entries by `SELECT ... WHERE "openstack_tag_production" = "true"` or similar. You cannot do this as you would need to specify the complete sets **in the correct order** for filtering.
Furthermore, my scheme allows the user to _drop_ certain tags using the telegraf machinery...

Reply:
Thank you @srebhan for clarifying.
I think we can query tags using regex in present config but yeah if additional tags are not a big issue then I'll implement tags as key-value pairs tags as keys and default value as 'true' and will also add tag_prefix and tag_value options.